exercises/practice/sgf-parsing/.meta/example.ex

```diff
@@ -0,0 +1,207 @@
+defmodule SgfParsing do
+  # Used to make recursive parsers lazy
+  defmacro lazy(parser) do
+    quote do
+      fn string -> unquote(parser).(string) end
+    end
+  end
+
+  defmodule Sgf do
+    defstruct properties: %{}, children: []
+  end
+
+  @type sgf :: %Sgf{properties: map, children: [sgf]}
+  @doc """
+  Parse a string into a Smart Game Format tree
+  """
+  @spec parse(encoded :: String.t()) :: {:ok, sgf} | {:error, String.t()}
+  def parse(encoded) do
+    parser = parse_tree_paren() |> eof()
+
+    with {:ok, tree, ""} <- run_parser(parser, encoded) do
+      {:ok, tree}
+    else
+      {:error, err, _rest} -> {:error, err}
+    end
+  end
+
+  # TREE PARSER
+
+  def parse_tree() do
+    lift2(
+      &%Sgf{properties: &1, children: &2},
+      char(?;)
+      |> error("tree with no nodes")
+      |> drop_and(many(parse_property()))
+      |> map(&Map.new/1),
+      lazy(
+        one_of([
+          map(parse_tree(), &List.wrap/1),
+          many(parse_tree_paren())
+        ])
+      )
+    )
+  end
+
+  def parse_tree_paren() do
+    char(?()
+    |> error("tree missing")
+    |> drop_and(parse_tree())
+    |> drop(char(?)))
+  end
+
+  def parse_property() do
+    lift2(
+      &{&1, &2},
+      some(satisfy(&(&1 not in '[();')))
+      |> map(&Enum.join(&1, ""))
+      |> validate(&(&1 == String.upcase(&1)), "property must be in uppercase"),
+      some(
+        char(?[)
+        |> error("properties without delimiter")
+        |> drop_and(many(escaped(&(&1 != ?]))))
+        |> drop(char(?]))
+        |> map(&Enum.join(&1, ""))
+      )
+    )
```

Comment:
I think it would be easier to read/understand this block if it were split apart,

```suggestion
    node_property = 
      some(satisfy(&(&1 not in '[();')))
      |> map(&Enum.join(&1, ""))
      |> validate(&(&1 == String.upcase(&1)), "property must be in uppercase")

    node_value = 
      some(
        char(?[)
        |> error("properties without delimiter")
        |> drop_and(many(escaped(&(&1 != ?]))))
        |> drop(char(?]))
        |> map(&Enum.join(&1, ""))
      )

    lift2(&{&1, &2}, node_property, node_value)
```